protobuf_serialization/writer.nim

```diff
@@ -1,236 +1,152 @@
-
-proc writeFixed(
-  stream: OutputStream,
-  fieldNum: int,
-  value: auto,
-  omittable: static bool
-) =
-  when sizeof(value) == 8:
-    let wire = Fixed64
-  else:
-    let wire = Fixed32
-  when omittable:
-    if value.unwrap() == 0:
-      return
-
-  stream.writeProtobufKey(fieldNum, wire)
-  stream.encodeFixed(value)
-
-proc writeValueInternal[T](stream: OutputStream, value: T)
-
-#stdlib types toProtobuf's. inlined as it needs access to the writeValue function.
-include stdlib_writers
-
-proc writeLengthDelimited[T](
-  stream: OutputStream,
-  fieldNum: int,
-  rootType: typedesc[T],
-  fieldName: static string,
-  flatValue: LengthDelimitedTypes,
-  omittable: static bool
-) =
-  const stdlib = type(flatValue).isStdlib()
-
-  var cursor = stream.delayVarSizeWrite(10)
-  let startPos = stream.pos
-
-  #Byte seqs.
-  when flatValue is CastableLengthDelimitedTypes:
-    if flatValue.len == 0:
-      cursor.finalWrite([])
-      return
-    stream.write(cast[seq[byte]](flatValue))
-
-  #Standard lib types which use custom converters, instead of encoding the literal Nim representation.
-  elif stdlib:
-    stream.stdlibToProtobuf(rootType, fieldName, fieldNum, flatValue)
-
-  #Nested object which even if the sub-value is empty, should be encoded as long as it exists.
-  elif rootType.isPotentiallyNull():
-    writeValueInternal(stream, flatValue)
-
-  #Object which should only be encoded if it has data.
-  elif flatValue is (object or tuple):
-    writeValueInternal(stream, flatValue)
-
-  else:
-    {.fatal: "Tried to write a Length Delimited type which wasn't actually Length Delimited.".}
-
-  const singleBuffer = type(flatValue).singleBufferable()
-  if (
-    (
-      #The underlying type of the standard library container is packable.
-      singleBuffer or (
-        #This is a object, not a seq or something converted to a seq (stdlib type).
-        (not stdlib) and (flatValue is (object or tuple))
-      )
-    ) and (
-      #The length changed, meaning this object is empty.
-      (stream.pos != startPos) or
-      #The object is empty, yet it exists, which is important as it can not exist.
-      rootType.isPotentiallyNull()
-    )
-  ):
-    cursor.finalWrite(newProtobufKey(fieldNum, LengthDelimited) & encodeVarInt(PInt(int32(stream.pos - startPos))))
-  else:
-    when omittable:
-      cursor.finalWrite([])
+import
+  std/typetraits,
+  stew/shims/macros,
+  faststreams/outputs,
+  serialization,
+  "."/[codec, internal, types]
+
+export outputs, serialization, codec, types
+
+proc writeValue*[T: object](stream: OutputStream, value: T)
+
+proc writeField(
+    stream: OutputStream, fieldNum: int, fieldVal: auto, ProtoType: type UnsupportedType) =
+  # TODO turn this into an extension point
+  unsupportedProtoType ProtoType.FieldType, ProtoType.RootType, ProtoType.fieldName
+
+proc writeField*[T: object](stream: OutputStream, fieldNum: int, fieldVal: T) =
+  # TODO https://github.com/status-im/nim-faststreams/pull/32
```

Comment:
Is the PR above fixing the only known issue with var-sized cursors? Why don't we merge it instead of introducing an inefficient work-around here?

Reply:
because the fix in faststtreams is dubious - when and if it gets resolved, this code can be updated, it doesn't materially matter - this PR, even with these copies here, does less  copying in general than the present version of the code. faststreams is its own nest of complexity and bugs, one thing at a time

Reply:
also, I'm considering an alternative approach here that precomputes the size of the fields instead of using cursors - since often, we want a _linear_ protobuf in memory, the approach with the cursors adds a lot of complexity and is actually slower because we have to copy _all_ data from the page buffers - it's likely faster to precompute the size of the full object, make one allocation and then write it - depends a bit on the streaming requirements